ampd/src/starknet/json_rpc.rs

```diff
@@ -87,48 +83,32 @@ where
     // `ContractCallEvent`
     async fn get_events_by_hash_contract_call(
         &self,
-        tx_hash: CheckedFelt,
-    ) -> Result<Vec<(FieldElementAndEventIndex, ContractCallEvent)>> {
+        message_id: FieldElementAndEventIndex,
+    ) -> Option<(FieldElementAndEventIndex, ContractCallEvent)> {
         let receipt_with_block_info = self
             .client
-            .get_transaction_receipt(tx_hash.clone())
+            .get_transaction_receipt(message_id.tx_hash.clone())
             .await
-            .map_err(StarknetClientError::FetchingReceipt)?;
+            .ok()?;
 
         if *receipt_with_block_info.receipt.execution_result() != ExecutionResult::Succeeded {
-            return Err(Report::new(StarknetClientError::UnsuccessfulTx));
+            return None;
         }
 
-        let mut message_id_and_event_pairs: Vec<(FieldElementAndEventIndex, ContractCallEvent)> =
-            vec![];
-
         match receipt_with_block_info.receipt {
             TransactionReceipt::Invoke(tx) => {
-                let mut event_index: u64 = 0;
-                // The RPC node always returns the events in the same order,
-                // so we can be sure that the msg ids will have the same indexes
-                // when iterating them
-                for e in tx.events.clone() {
-                    if let Ok(cce) = ContractCallEvent::try_from(e.clone()) {
-                        let message_id =
-                            FieldElementAndEventIndex::new(tx_hash.clone(), event_index)
-                                .map_err(StarknetClientError::MessageIdConstruction)?;
-                        message_id_and_event_pairs.push((message_id, cce));
-                    }
-                    // we iterate regardless of the event type,
-                    // because that's what we do in the relayer
-                    event_index = event_index
-                        .checked_add(1)
-                        .ok_or(StarknetClientError::OverflowingU64)?;
-                }
-            }
-            TransactionReceipt::L1Handler(_) => (),
-            TransactionReceipt::Declare(_) => (),
-            TransactionReceipt::Deploy(_) => (),
-            TransactionReceipt::DeployAccount(_) => (),
-        };
+                let event_index: usize = message_id.event_index.try_into().ok()?;
+                let event = tx.events.get(event_index)?;
 
-        Ok(message_id_and_event_pairs)
+                ContractCallEvent::try_from(event.clone())
+                    .map(|cce| (message_id, cce))
+                    .ok()
+            }
+            TransactionReceipt::L1Handler(_) => None,
+            TransactionReceipt::Declare(_) => None,
+            TransactionReceipt::Deploy(_) => None,
+            TransactionReceipt::DeployAccount(_) => None,
```

Comment:
why not `_ => None`

Reply:
AFAIK it's a best practice to exhaust all variants. Main reason is that they might add a new `TransactionReceipt` type, which we might want to handle. Applies for other enums, coming from libs, in general.
Apart from that, no other reason. 
I'll do `_ => None` if you prefer.